### PR TITLE
[FLINK-35351][checkpoint] Fix fail during restore from unaligned checkpoint with custom partitioner

### PR DESCRIPTION
**Pull request by pnowojski**
The same as https://github.com/apache/flink/pull/24784 but with reworked testing code

## What is the purpose of the change

To fix [FLINK-35351](https://issues.apache.org/jira/browse/FLINK-35351)

## Verifying this change

UnalignedCheckpointRescaleITCase Topology.CUSTOM_PARTITIONER

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? ( no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointRescaleITCase.java [hunk 1]
```diff
@@ -336,38 +335,57 @@ public void create(
...
 
             @Override
             public void create(
-                    StreamExecutionEnvironment environment,
+                    StreamExecutionEnvironment env,
                     int minCheckpoints,
                     boolean slotSharing,
-                    int expectedFailuresUntilSourceFinishes,
+                    int expectedRestarts,
                     long sourceSleepMs) {
```
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointRescaleITCase.java:345]: The old version was silently ignoring `minCheckpoints`, `expectedRestarts` and `sourceSleepMs` parameters, without throwing any kind of error.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointRescaleITCase.java [hunk 2]
```diff
@@ -761,43 +778,12 @@ public int partition(String key, int numPartitions) {
...
-        }
-
+    private static class BackPressureInducingSink<T> implements SinkFunction<T> {
         @Override
-        public void initializeState(FunctionInitializationContext context) {
-            // do  nothing
+        public void invoke(T value, Context ctx) throws Exception {
+            // TODO: maybe similarly to VerifyingSink, we should back pressure only until some point
+            // but currently it doesn't seem to be needed (test runs quickly enough)
+            Thread.sleep(1);
```
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointRescaleITCase.java:786]: It looks like the simple version is good enough.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointRescaleITCase.java [hunk 3]
```diff
@@ -336,38 +335,57 @@ public void create(
...
+                                new LongSource(
+                                        minCheckpoints,
+                                        parallelism,
+                                        expectedRestarts,
+                                        env.getCheckpointInterval(),
+                                        sourceSleepMs),
+                                noWatermarks(),
+                                "source")
                         .name("source")
+                        .uid("source")
```
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointRescaleITCase.java:358]: old version was missing `uid`, which could cause problems after rescaling

**Review by StefanRRichter — APPROVED** [review]

**Comment by ldadima** [comment]
Thanks for changes. LGTM

**Review by ldadima — APPROVED** [review]

**Pull request merged**